src/icrar/leap-accelerate/algorithm/cuda/PhaseRotate.cu

```diff
@@ -67,65 +70,139 @@ namespace cuda
-        Integration& integration,
-        MetaDataCudaHost& metadata,
-        const MVDirection& direction)
+    __device__ __forceinline__ cuDoubleComplex cuCexp(cuDoubleComplex z)
     {
-        //using namespace std::literals::complex_literals;
-        auto& data = integration.data;
-        auto& uvw = integration.uvw;
-        auto parameters = integration.parameters;
+        // see https://forums.developer.nvidia.com/t/complex-number-exponential-function/24696/2
+        cuDoubleComplex res;
+        sincos(z.y, &res.y, &res.x);
+        double t = exp(z.x);
+        res.x *= t;
+        res.y *= t;
+        return res;
+    }
 
-        if(metadata.init)
-        {
-            //metadata['nbaseline']=metadata['stations']*(metadata['stations']-1)/2
-            
-            metadata.SetDD(direction);
-            metadata.SetWv();
-            // Zero a vector for averaging in time and freq
-            metadata.avg_data = Eigen::MatrixXcd(integration.baselines, metadata.GetConstants().num_pols);
-            metadata.init = false;
-        }
-        metadata.CalcUVW(uvw);
+    __global__ void g_RotateVisibilities(
+        //Integration integration,
+        Constants constants,
+        Eigen::Matrix3d dd,
+        double2 direction,
+        double3* uvw, int uvwLength,
+        double3* oldUVW, int oldUVWLegth,
+        cuDoubleComplex* pavg_data, int avg_dataRows, int avg_dataCols)
+    {
+        using VectorXcucd = Eigen::Matrix<cuDoubleComplex, Eigen::Dynamic, 1>;
+        using MatrixXcucd = Eigen::Matrix<cuDoubleComplex, Eigen::Dynamic, Eigen::Dynamic>;
+        using Tensor3Xcucd = Eigen::Matrix<VectorXcucd, Eigen::Dynamic, Eigen::Dynamic>;
+
+        int integration_baselines = 0;
+        Tensor3Xcucd integration_data; //TODO: incomplete
+        Eigen::Map<MatrixXcucd> avg_data = Eigen::Map<MatrixXcucd>(pavg_data, avg_dataRows, avg_dataCols);
 
-        // loop over baselines
-        for(int baseline = 0; baseline < integration.baselines; ++baseline)
+        /// loop over baselines
+        for(int baseline = 0; baseline < integration_baselines; ++baseline)
         {
-            // For baseline
-            const double pi = boost::math::constants::pi<double>();
-            double shiftFactor = -2 * pi * uvw[baseline].get()[2] - metadata.oldUVW[baseline].get()[2]; // check these are correct
-            shiftFactor = shiftFactor + 2 * pi * (metadata.GetConstants().phase_centre_ra_rad * metadata.oldUVW[baseline].get()[0]);
-            shiftFactor = shiftFactor -2 * pi * (direction.get()[0] * uvw[baseline].get()[0] - direction.get()[1] * uvw[baseline].get()[1]);
+            const double pi = CUDART_PI;
+            double shiftFactor = -2 * pi * uvw[baseline].z - oldUVW[baseline].z;
+            shiftFactor = shiftFactor + 2 * pi * (constants.phase_centre_ra_rad * oldUVW[baseline].x);
+            shiftFactor = shiftFactor - 2 * pi * (direction.x * uvw[baseline].x - direction.y * uvw[baseline].y);
```

Comment:
`shiftFactor += ...` and `shiftFactor -= ...` maybe?

Reply:
And maybe it's cheaper also to factor out `2 * pi`, do all additions/subtractions first, then finally multiply by `2 * pi` once only.

Reply:
I recall I ported parts of this wrong too. I'll update with the corrections